src/software/ai/hl/stp/tactic/goalie_tactic.cpp

```diff
@@ -152,155 +153,161 @@ void GoalieTactic::calculateNextAction(ActionCoroutine::push_type &yield)
-            {
-                if (ball.position().y() > 0)
-                {
-                    goalie_pos =
-                        field.friendlyGoalpostPos() + Vector(0, -ROBOT_MAX_RADIUS_METERS);
-                }
-                else
-                {
-                    goalie_pos =
-                        field.friendlyGoalpostNeg() + Vector(0, ROBOT_MAX_RADIUS_METERS);
-                }
-            }
-            else
-            {
-                goalie_pos = *clamped_goalie_pos;
-            }
-            Angle goalie_orientation = (ball.position() - goalie_pos).orientation();
-
-            move_action->updateControlParams(*robot, goalie_pos, goalie_orientation,
-                                             goalie_final_speed, DribblerEnable::OFF,
-                                             MoveType::NORMAL, AutochickType::AUTOCHIP,
-                                             BallCollisionType::ALLOW);
-            next_action = move_action;
+            next_action = positionToBlockShot(move_action);
         }
 
         yield(next_action);
     } while (!move_action->done());
 }
 
+std::vector<Point> GoalieTactic::getIntersectionsBetweenBallVelocityAndFullGoalSegment()
+{
+    // compute intersection points from ball position and velocity
+    Ray ball_ray = Ray(ball.position(), ball.velocity());
+
+    // Create a segment along the goal line, slightly shortened to account for the
+    // robot radius so as we move along the segment we don't try to run into the goal
+    // posts. This will be used in case 3 as a fallback when we don't have an
+    // intersection with the crease lines
+    Segment full_goal_segment =
+        Segment(field.friendlyGoalpostNeg() + Vector(0, -ROBOT_MAX_RADIUS_METERS),
+                field.friendlyGoalpostPos() + Vector(0, ROBOT_MAX_RADIUS_METERS));
+
+    return intersection(ball_ray, full_goal_segment);
+}
+
+std::shared_ptr<Action> GoalieTactic::panicAndStopBall(
+    std::shared_ptr<MoveAction> move_action, const Point &stop_ball_point)
+{
+    // the ball is heading towards the net, move to intercept the shot
+    // the final speed is a dynamic parameter so that if the goalie needs
+    // to dive for the shot instead of stop when reaching the intersection
+    // point it can do so.
+    Point goalie_pos =
+        closestPoint((*robot).position(), Segment(ball.position(), stop_ball_point));
+    Angle goalie_orientation = (ball.position() - goalie_pos).orientation();
+
+    move_action->updateControlParams(*robot, goalie_pos, goalie_orientation, 0.0,
+                                     DribblerEnable::OFF, MoveType::NORMAL,
+                                     AutochickType::AUTOCHIP, BallCollisionType::ALLOW);
+    return move_action;
+}
+
+std::shared_ptr<Action> GoalieTactic::chipBallIfSafe(
+    std::shared_ptr<ChipAction> chip_action, std::shared_ptr<StopAction> stop_action)
+{
+    // if the ball is in the don't chip rectangle we do not chip the ball
```

Comment:
"if the ball is in the __" seems to be missing a word.

Reply:
clarified